tgstation/tgstation
# Ordnance Content Update: Scientific Papers

## How do I play/test/operate this?
1.  Download NT Frontier on any modular computers. It should debrief you on what experiments are available and how to publish.
2.  If you want to do a bomb experiment, make sure it's captured by the doppler array (as usual) and then print the experiments into a disk and publish it. 
3. If you want to do a gas experiment, make the gas and either pump it into a tank and 1) overpressurize it with a "clear" gas like N2 or 2) overpressurize tanks with the gas itself. Make sure you do the overpressurizing in the compressor machine. When tanks are destroyed/ejected leaked gas will get recorded. Print it into a disk and publish it.
4. For publication, the file needs to be directly present **inside** the computer's HDD. This means you need to copy it first with the file manager.
5. Fill the data (if desired, it will autofill with boiler plate if you dont) and send away!
6. Doing experiments unlock nodes, while doing them well unlocks boosts (which are discounts but slightly more restrictive) which are purchaseable with NT Frontier.
7. If you are testing this and have access to admin tools, there are various premade bombs under `obj/effect/spawner/newbomb`

## About The Pull Request
A doc I wrote detailing the why and what part of this PR.
https://hackmd.io/JOakSYVMSh2zU2YL5ju_-Q
TLDR: Added new experiment to toxins, added the framework for those experiments existing. Experiments comes in gas synthesis and also bombs but with more parameters. Experiments needs to be published through papers, various choices to be made there.

- [x] Combine implosion compressor and bomb processor into one.
- [x] Make the UI up to standard.
- [x] Add a sprite for the compressor.
- [x] Map changes.

<details>
<summary> Codersprites </summary>
They're not very good but they're serviceable at least.
<br/>
<image src="https://user-images.githubusercontent.com/54709710/141190268-f88eba07-a2f7-4c1f-8943-7140424ac948.png">
<br/>
<image src="https://user-images.githubusercontent.com/54709710/141190440-bf26a4da-829d-49f5-a4aa-983910dcff90.png">
<br/>
<image src="https://user-images.githubusercontent.com/54709710/141189632-e9d5f875-7ed4-4948-8f3f-2226abb52d95.png">
<image src="https://user-images.githubusercontent.com/54709710/141189743-f4cdbaf2-d3c9-4b43-9ece-91a1c51e0e60.png">
<image src="https://user-images.githubusercontent.com/54709710/141189892-aa42c875-32db-41b3-9c49-d04f1a26b7b3.png">
</details>

<details>
<summary> Doppler malf ban he </summary>
<img src="https://user-images.githubusercontent.com/54709710/144702283-123d2064-d054-4498-a4a6-a688f72416f5.png"><br/>
small change but i couldn't resist making the joke
</details>

Implementation notes:
Because of how paper works, ordnance experiments are handled outside of experiment_handler components. My reasoning for this is twofold: 
1) The experiments will be completed manually on publication and if the experiment isn't unlocked yet it will still be completed. 
2) Experiment handler datums have several procs which require an atom-level parent, and I figured this is the most sensible and cleanest way to implement this without changing the experiment handler datum too much. 
I won't mind changing this if requested by a maintainer.

Small change to `/obj/machinery/proc/power_change()` signal ordering to adjust the state first and then send the signal. Didn't found any other usage of this signal except mine but barge down my door if it broke something.

Rewrote the ttv merge_gases() code to be slightly more readable.
A small code improvement for thermomachine to use tofixed (my fault).

## Why It's Good For The Game
Adds content to ordnance, more experiments, more variety. 
More sensible learning curve, better player feedback.
Pretty much refer to my hackmd for a more substantial justification.

## Changelog
:cl:
add: Ordnance have been updated to enable the publication of papers
add: Several new explosive and gas synthesis experiments have been added to ordnance
add: Anomaly compressor has been TGUIzed and now supports simulating the reaction of the gases inside the ttv.
add: New tank compressor machine for toxins. You can overpressurize tanks with exotic gases and complete experiments.
balance: Several techweb nodes are locked and require toxin experiments to complete.
balance: Toxins can purchase boosts for various techweb nodes.
balance: You no longer need to anchor doppler arrays for it to work.
qol: Doppler array and implosion compressor now supports deconstruction, implosion compressor construction added.
qol: Doppler now emits a red light to denote it's direction and it being on. Doppler not malf.
code: Implosion compressor renamed to anomaly refinery.
code: Created a new program tab "Science" for the downloader app. Removed Robotics.
refactor: Reworked the code for bombspawner (used in the cuban pete arcade game)
/:cl:

## UI:
<details>
<summary>Tank Compressor</summary>
<img src="https://user-images.githubusercontent.com/54709710/141077933-5556c9d1-21fc-406a-864b-58cd4c7f1cef.png">
<img src="https://user-images.githubusercontent.com/54709710/141078004-66d448f3-a4c9-49aa-b0e6-84c4db76c6d8.png">
</details>

<details>
<summary>Publisher</summary>
<img src="https://user-images.githubusercontent.com/54709710/141078351-b567a13a-a0d7-4004-ae1c-89fa8f33b5c8.png">
<img src="https://user-images.githubusercontent.com/54709710/141078432-5f84ee8d-a8e9-45a0-876f-d84103449172.png">
<img src="https://user-images.githubusercontent.com/54709710/141078581-72e86cb6-c30f-452d-a70a-7eb379db5086.png">
<img src="https://user-images.githubusercontent.com/54709710/141078512-6fd20566-b2f8-4780-98a3-eea32f9822af.png">
</details>

<details>
<summary>Doppler</summary>
<img src="https://user-images.githubusercontent.com/54709710/138489770-a7ef5d45-cf3e-48ac-8f63-ee5ae487e6e6.png">
</details>

<details>
<summary>Detonation Chamber / Anomaly Compressor</summary>
<img src="https://user-images.githubusercontent.com/54709710/141079055-5082acc5-26af-4211-8e74-db9c13fe6291.png">
<img src="https://user-images.githubusercontent.com/54709710/141078817-8deb5fd0-e32b-4a88-a726-8f6628435be7.png">
</details>
